allensdk/brain_observatory/pickle_utils.py

```diff
@@ -0,0 +1,149 @@
+from typing import Union, Any
+import pickle
+import gzip
+import pathlib
+
+
+def load_and_sanitize_pickle(
+        pickle_path: Union[str, pathlib.Path]) -> Any:
+    """
+    Load the data from a pickle file and pass it through
+    sanitize_pickle_data, so that all bytes in the data are
+    cast to strings.
+
+    Parameters
+    ----------
+    pickle_path: Union[str, pathlib.Path]
+        Path to the pickle file
+
+    Returns
+    -------
+    pickle_data: Any
+        The data that was in the pickle file
+
+    Notes
+    -----
+    Because sanitize_pickle_data alters the data in-place,
+    this method encapsulates loading and sanitizing so that
+    users do not think they have a pre-sanitization copy
+    of the data available.
+    """
+    if isinstance(pickle_path, str):
+        pickle_path = pathlib.Path(pickle_path)
+
+    if pickle_path.name.endswith('gz'):
+        open_method = gzip.open
+    elif pickle_path.name.endswith('pkl'):
+        open_method = open
+    else:
+        raise ValueError("Can open .pkl and .gz files; "
+                         f"you gave {pickle_path.resolve().absolute()}")
+
+    with open_method(pickle_path, 'rb') as in_file:
+        raw_data = pickle.load(in_file, encoding='bytes')
+    return _sanitize_pickle_data(raw_data)
+
+
+def _sanitize_pickle_data(
+        raw_data: Union[list, dict]) -> Union[list, dict]:
+    """
+    Sometimes data read from the pickle file comes with keys that
+    are strings; sometimes it comes with keys that are bytes.
+
+    This method iterates over the elements in the pickle file, casting
+    the bytes to strings, returning the same object with the mapped
+    keys.
+
+    Note
+    ----
+    Alters raw_data in-place
+    """
+    if isinstance(raw_data, dict):
+        raw_data = _sanitize_dict(raw_data)
+    elif isinstance(raw_data, list):
+        raw_data = _sanitize_list(raw_data)
+
+    return raw_data
+
+
+def _sanitize_list(
+        raw_data: list) -> list:
+    """
+    Sanitize a list read from the pickle file, casting bytes
+    into str and returning the sanitized list.
+
+    Note
+    ----
+    Alters raw_data in place
+    """
+    for idx, element in enumerate(raw_data):
+        if isinstance(element, list) or isinstance(element, tuple):
+            raw_data[idx] = _sanitize_list_or_tuple(element)
+        elif isinstance(element, dict):
+            raw_data[idx] = _sanitize_dict(element)
+        elif isinstance(element, bytes):
+            raw_data[idx] = element.decode('utf-8')
+
+    return raw_data
```

Comment:
Do you want to add an `else` block here?

Reply:
If you make it through all those `elif`s, then the correct thing to do is nothing. I can add a `pass`, though.